consultation_analyser/consultations/models.py

```diff
@@ -40,301 +41,237 @@ class Meta:
+        if self.pk:
+            slug_exists = (
+                ModelClass.objects.filter(slug=generated_slug).exclude(pk=self.pk).exists()
+            )
+        else:
+            slug_exists = ModelClass.objects.filter(slug=generated_slug).exists()
+        timestamp = datetime.datetime.now().strftime("%Y%m%d-%H%M%S%f")
+        if not generated_slug:
+            generated_slug = timestamp
+        elif slug_exists:
+            generated_slug = f"{generated_slug}-{timestamp}"
+        self.slug = generated_slug
+        return super().save(*args, **kwargs)
 
-    def delete(self, *args, **kwargs):
-        """
-        Delete Related theme objects that can become orphans
-        because deletes will not cascade from Answer because
-        the Theme in question could still be associated with
-        another Answer
-        """
-        Theme.objects.filter(answer__question__section__consultation=self).delete()
+    class Meta:
+        abstract = True
 
-        super().delete(*args, **kwargs)
 
-    @property
-    def latest_processing_run(self):
-        processing_runs = ProcessingRun.objects.filter(consultation=self).order_by("created_at")
-        latest = processing_runs.last() if processing_runs else None
-        return latest
+class Consultation(UUIDPrimaryKeyModel, TimeStampedModel, SlugFromTextModel):
+    users = models.ManyToManyField(User)
 
-    def get_processing_run(self, processing_run_slug=None):
-        # If None, get latest run if exists
-        if processing_run_slug:
-            processing_run = ProcessingRun.objects.get(consultation=self, slug=processing_run_slug)
-        else:
-            processing_run = self.latest_processing_run
-        return processing_run
+    class Meta(UUIDPrimaryKeyModel.Meta, TimeStampedModel.Meta, SlugFromTextModel.Meta):
+        constraints = [
+            models.UniqueConstraint(fields=["slug"], name="unique_consult_slug"),
+        ]
 
 
-class Section(UUIDPrimaryKeyModel, TimeStampedModel):
+class Question(UUIDPrimaryKeyModel, TimeStampedModel, SlugFromTextModel):
+    text = models.TextField()
     consultation = models.ForeignKey(Consultation, on_delete=models.CASCADE)
-    name = models.TextField()
-    slug = models.CharField(null=False, max_length=256)
+    order = models.IntegerField(null=True)
 
-    class Meta:
+    class Meta(UUIDPrimaryKeyModel.Meta, TimeStampedModel.Meta, SlugFromTextModel.Meta):
         constraints = [
-            models.UniqueConstraint(
-                fields=["slug", "consultation"], name="unique_section_consultation"
-            ),
+            models.UniqueConstraint(fields=["slug"], name="unique_q_slug"),
         ]
 
 
-class MultipleChoiceNotProportionalError(Exception):
-    pass
-
-
-@dataclass
-class MultipleChoiceQuestionStats:
-    question: str
-    counts: dict[str, int]
-    total_responses: int
-    has_multiple_selections: bool
+class QuestionPart(UUIDPrimaryKeyModel, TimeStampedModel):
```

Comment:
Given that there are a few places where we're assuming there's only one free text type per question, do we want to include some model validation to confirm this?

Reply:
Yes - that's a good idea.

Reply:
Uniqueness constraint (at most one free-text part per question) added: https://github.com/i-dot-ai/consult/pull/405/commits/517f14391a6c9008a953ce9152e36ae3e20adb92.